[stdlib] DoubleWidth Implementation

natecook1000 opened the pull request:
cc @moiseev

natecook1000 committed:
[stdlib] Partial implementation of DoubleWidth

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3247,208 +3294,348 @@ public struct DoubleWidth<T : FixedWidthInteger>
+
+  public func quotientAndRemainder(dividingBy other: DoubleWidth)
+    -> (quotient: DoubleWidth, remainder: DoubleWidth) {
+    let isNegative = (self < DoubleWidth()) != (other < DoubleWidth())
```
We can't use `0` here?

natecook1000 replied:
I think this was from before I implemented the integer literal init—works now for values that aren't larger than an `Int`.

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3247,208 +3294,348 @@ public struct DoubleWidth<T : FixedWidthInteger>
 
-  public var littleEndian: DoubleWidth<T> {
-    fatalError()
+  public var littleEndian: DoubleWidth {
```
Is the code for the methods that use `byteSwapped` all boilerplate that gets repeated for all integer types, and if so can we factor it out (e.g. into a default implementation)?

natecook1000 replied:
It is! Factored that out into a `FixedWidthInteger` default implementation—`byteSwapped` is the customization point. I didn't remove the implementations from the standard integer types.

natecook1000 replied:
> Why not?

Compelling argument! 😂 Those are out now too; tests updated.

xwu replied:
Are there not performance implications to losing the concrete implementations?

natecook1000 replied:
I would think specialization would take care of that, but I could be wrong…

dabrahams replied:
You could write a benchmark, but looking at generated code is probably enough.

natecook1000 committed:
Use literals instead of DoubleWidth inits

natecook1000 committed:
Convert endian symbols to default implementations

natecook1000 committed:
Add negation operator/method to DoubleWidth

natecook1000 committed:
A few tests for DoubleWidth

dabrahams commented:
on Sun May 07 2017, Nate Cook <notifications-AT-github.com> wrote:

> I didn't
> remove the implementations from the standard integer types.

Why not?

natecook1000 committed:
Provide implementations or masking bitshift operators

natecook1000 committed:
Remove endian operations from fixed-width ints

natecook1000 committed:
Add recursive constraints for integer Magnitude

natecook1000 committed:
Update tests for endian default implementations

natecook1000 committed:
Fix overflow issues

natecook1000 committed:
Slightly better DoubleWidth testing

natecook1000 committed:
Implement tricky full-width multiply/divides

natecook1000 committed:
Fix overflow issue

natecook1000 committed:
Shortcut on binary division

natecook1000 committed:
More tests for DoubleWidth

natecook1000 committed:
Add floating-point initializers and tests

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Revert multipliedFullWidth for 64-bit ints on 32-bit platforms

natecook1000 commented:
@swift-ci Please smoke test

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
       _storage.high.${x.name}ReportingOverflow(rhs._storage.high)
     let isLowOverflow = lowOverflow == .overflow
-    let result = (high ${x.operator} (isLowOverflow ? 1 : 0), low)
+    let result = (high &${x.operator} (isLowOverflow ? High(1) : High(0)), low)
```
`High` conforms to `ExpressibleByIntegerLiteral`, and you removed the explicit usage of types in favor of just literals elsewhere =)

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    let initialOffset = q.leadingZeroBitCount + (DoubleWidth.bitWidth - rhs.leadingZeroBitCount) - 1
+
+    // Start with remainder capturing the high bits of q.
+    var r = q >> Magnitude(DoubleWidth.bitWidth - initialOffset)
```
It would be interesting to have some benchmarks and the effect of using `>>` vs `&>>` here.

natecook1000 replied:
Adding a note about that.

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
-    /*_log("(II) 4 (\(high), \(low)) carry: \(carry)")*/
+  public func dividedReportingOverflow(by other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 || (other < 0 && self == .min) {
```
Shouldn't it be `|| (other == -1 && self == .min)`?

natecook1000 replied:
Need to check for signedness, but 👍

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
-  }
+  public func remainderReportingOverflow(dividingBy other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 {
```
Shouldn't remainder overflow in case of `.min / -1` as well?

natecook1000 replied:
This is funny, because the remainder (0) is representable, but the quotient is not.

moiseev replied:
I now remember the discussion about integers whether the remainder should overflow in this case, and I believe we ended up agreeing that it should not.
Regardless, the behavior should be consistent with that of built-in integer types.

moiseev replied:
Huh! It is a compiler error =) So I guess we stick to the plan. Overflow that is.
```swift
(swift) Int8.min.remainderReportingOverflow(dividingBy: -1 as Int8)
<REPL Input>:1:10: error: division '-128 % -1' results in an overflow
Int8.min.remainderReportingOverflow(dividingBy: -1 as Int8)
         ^
```

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
 % end
 
+  public static func <<=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
+    if rhs < DoubleWidth((0, 0)) {
```
`DoubleWidth` should also conform to `ExpressibleByIntegerLiteral` right? So this can be replaced by just `rhs < (0 as DoubleWidth)`? Not a big win in readability, but at least no `((...))` ;-)

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
 
+  public static func <<=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
+    if rhs < DoubleWidth((0, 0)) {
+      lhs >>= DoubleWidth((0, 0)) - rhs
```
Prefix `-` should be able to take care of this as well... I think..

natecook1000 replied:
Prefix `-` is only available for signed `DoubleWidth`s, so I'm subtracting from zero in non-constrained methods.

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    
+    // Shift is exactly the width of `Base`, so low -> high.
+    if Int(rhs._storage.low) == Base.bitWidth {
+      lhs = DoubleWidth((High(extendingOrTruncating: lhs._storage.low), numericCast(0)))
```
`numericCast` should not be necessary. Without it the type context is unambiguous, and literal would get the right type... Well, in ideal world, I mean.

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
   public init(_builtinIntegerLiteral x: _MaxBuiltinIntegerType) {
-    fatalError("Method must be overridden")
+    // FIXME: This won't work if `x` is out of range for `Int`
+    self.init(Int(_builtinIntegerLiteral: x))
```
Will it change anything if we use `Int64` explicitly? I mean, it's not perfect still, but at least on 32-bit platforms it would support larger literals.

moiseev reviewed (commented):
Just a thought... Now that the `DoubleWidth<T>` has implementation, it is possible and quite interesting to create a 2- or 3-bit `FixedWidthInteger` and test `DoubleWidth<T>` exhaustively.

moiseev commented:
Overall @natecook1000 this is AMAZING! Thanks a ton! :+1:

natecook1000 committed:
Catch overflow into high bit in signed multiplication

natecook1000 commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -1339,7 +1339,11 @@ extension Numeric {
+  where Magnitude : BinaryInteger, Magnitude.Magnitude == Magnitude
+{
+  // FIXME(ABI) (Recursive Protocol Constraints): should add:
+  // where Magnitude : BinaryInteger
```
Yep 🙄

jrose-apple commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
-  public init<T : BinaryInteger>(extendingOrTruncating source: T) {
-    fatalError()
+    // Can't represent a negative 'source' if Base is unsigned
+    guard source >= 0 || DoubleWidth<Base>.isSigned else { return nil }
```
Having to write `0 as T` is terrible for real-world code. I think this should be taken as more evidence that the mixed-type operators are not a good idea with our current set of overload rules.

moiseev replied:
Nobody has to write `(0 as T)`. It will work without an explicit type context just fine, unless you're messing with the guts of integers, where incorrect overload can lead to infinite recursion.

natecook1000 committed:
Incorporate feedback from @moiseev

- Also clean up some 80-column issues
- And improve some tests from before literal expressibility

natecook1000 committed:
Add an abstract for DoubleWidth

natecook1000 commented:
A three- or five-bit integer would be particularly good for testing, since there could easily be power-of-two assumptions in the code. And like, does a type's `Magnitude` need to have the same bit width as the type? Technically, all the signed types we have could just use `UInt64`.

moiseev commented:
@natecook1000 five bits is too much for exhaustive testing. 3 is just right, I think. Also `DoubleWidth` uses `Magnitude` for `Low`, so using `UInt64` for everything is probably not the best idea, even though possible.

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3232,44 +3182,65 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    } else {
+      // At this point we know source's bitWidth > Base.bitWidth, or else
+      // we would've taken the first branch.
+      let lowInT = source & T(~Low(0))
```
```swift
let lowInT = source & T(~(0 as Low))
```

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    // FIXME: spurious compile error when 'where' clause above is removed:
+    // error: non-nominal type 'T.RawSignificand' does not support explicit initialization
+    let fractionPart: Raw = bitPattern &<< Raw(Raw.bitWidth - offset)
+    guard fractionPart == 0 else {
```
```swift
guard fractionPart == (0 as Raw) else { ... }
```

dabrahams replied:
zero is zero is zero.  I don't see any point in being explicit about the type here.

xwu replied:
Of course, it is guaranteed to yield the same result. The idea is that the heterogeneous comparison with `0 as IntegerLiteralType` is less performant, although @moiseev demonstrates that any such loss is small.

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    let highBit = ~(~0 >> 1) as Magnitude
+    for _ in initialOffset..<DoubleWidth.bitWidth {
+      r <<= 1
+      if q & highBit != 0 {
```
```swift
if q & highBit != (0 as Magnitude) { ... }
```

natecook1000 commented:
Right, was just thinking that's an undocumented assumption of `Magnitude`—should we specify that as a semantic requirement of `FixedWidthInteger` even though the type system can't enforce it? Are there cases where you'd want a fixed-width integer with a different size magnitude? The way things are structured it kind of means "the unsigned version of this type," but I don't think we actually state that anywhere. @dabrahams

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
-    /*_log("(II) 4 (\(high), \(low)) carry: \(carry)")*/
+  public func dividedReportingOverflow(by other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 || (DoubleWidth.isSigned && other == -1 && self == .min) {
```
Here, do you intend a `DoubleWidth` to `Int` comparison when you write `other == 0` and `other == -1`?

natecook1000 replied:
`other == 0` should be same type, but `other == -1` is maybe safer as an `Int` comparison. The signed check should make it safe, but it worries me that the -1 literal could get converted into an unsigned `DoubleWidth`.

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
-  }
+  public func remainderReportingOverflow(dividingBy other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 || (DoubleWidth.isSigned && other == -1 && self == .min) {
```
Ditto here.

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    func sum(_ x: Low, _ y: Low, _ z: Low) -> (partial: Low, carry: Low) {
+      let (sum1, overflow1) = x.addingReportingOverflow(y)
+      let (sum2, overflow2) = sum1.addingReportingOverflow(z)
+      let carry: Low = (overflow1 == .overflow ? Low(1) : Low(0))
```
No need for `Low(1)` and `Low(0)` here because you've written `let carry: Low`.

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+      return
+    }
+    
+    if rhs._storage.high != 0 || rhs._storage.low >= DoubleWidth.bitWidth {
```
```swift
if rhs._storage.high != (0 as High) || ...
```

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    }
+
+    // Shift is larger than this type's bit width.
+    if rhs._storage.high != 0 || rhs._storage.low >= DoubleWidth.bitWidth {
```
```swift
if rhs._storage.high != (0 as High) || ...
```

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
 
   public var leadingZeroBitCount: Int {
-    fatalError()
+    return high == 0
```
```swift
return high == (0 as High) ? ... : ...
```

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
 
   public var trailingZeroBitCount: Int {
-    fatalError()
+    return low == 0
```
```swift
return low == (0 as Low) ? ... : ...
```

xwu reviewed (commented):
A few more heterogeneous comparison issues.

I think, after having experimented with #9932, the `0xFFFF_FFFF_FFFF_FFFF` scenario isn't a big deal because the diagnostics are good, but it _is_ annoying that for optimal performance one must now always specify the type, but only for comparison and not for arithmetic. I haven't thought very deeply about whether the same issue with performance applies to heterogeneous bit shifts.

I do wonder if for now the best idea would be not to ship the heterogeneous operators (edit: or, to ship them in a standalone protocol to which only concrete types conform).

Edit: FWIW, this is a practical look at the impact for generic code, which I've just updated to avoid unnecessary heterogeneous comparison: https://github.com/xwu/NumericAnnex/pull/3/commits/042ae884abecfabc9f655041da8eea7d58854bfd

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    let result = DoubleWidth(extendingOrTruncating: product)
+    
+    let isNegative = (self < (0 as DoubleWidth)) != (rhs < (0 as DoubleWidth))
+    let didCarry = isNegative ? carry != ~0 : carry != 0
```
Yikes, this is where the heterogeneous operators really start biting. Presumably, you didn't mean `~0` as in `Int(-1)`.

```swift
let didCarry = isNegative ? carry != ~(0 as DoubleWidth) : carry != (0 as DoubleWidth)
```

natecook1000 replied:
Yikes, thanks for finding this!

moiseev commented:
@xwu FWIW, we did not modify our benchmark code at all when the new integers landed. So if there was any degradation in performance it was there and was deemed acceptable. I remember playing with explicit type context in various places, and measuring difference in compilation time as well as performance of the resulting code, and not seeing much difference.

The last point you made, though, is really frightening. I'm talking about `x != ~0`. Testing it now.

moiseev commented:
```swift
func f<T : BinaryInteger>(_ x: T) -> Bool {
  return x != ~0
}

func g<T : BinaryInteger>(_ x: T) -> Bool {
  return x != (~0 as T)
}

print(f(UInt.max))
print(g(UInt.max))
```

Outputs:
```
true
false
```

And this is bad. 😞

moiseev commented:
```swift
@inline(never)
func f<T : BinaryInteger>(_ x: T) -> Bool {
  return x != 0
}

@inline(never)
func g<T : BinaryInteger>(_ x: T) -> Bool {
  return x != (0 as T)
}

public var result = false

for i 0 ..< 10_000_000_000 {
  result = result || f(i)
  // result = result || g(i)
}
```

Rough performance measurements of this code for `f` and `g` are virtually the same (8.062 for `f` vs 8.014 for `g`).

dabrahams commented:
So, this is not simply a problem with heterogeneous comparison, but an interaction with the way literals and overloading work.

@DougGregor We may need to have that discussion about literals sooner than we thought.

dabrahams commented:
@natecook1000 Regarding https://github.com/apple/swift/pull/9367#issuecomment-304154500 (GH doesn't give me a place to reply), I can't see enough context to guess what assumption you're talking about.

dabrahams commented:
on Thu May 25 2017, Nate Cook <notifications-AT-github.com> wrote:

> A three- or five-bit integer would be particularly good for testing,
> since there could easily be power-of-two assumptions in the code. And
> like, does a type's `Magnitude` need to have the same bit width as the
> type? 

OK, I see now what you were referring to.  No, a type's Magnitude
doesn't need to have the same bit width, but as Max wrote, it's a good
idea to use a Magnitude of minimal width.

-- 
-Dave

natecook1000 commented:
The assumption is that a fixed width integer's `Magnitude` will always have the same `bitWidth`as the type itself. The assumption is explicit with `DoubleWidth`, since it uses `Magnitude` as the lower half.

dabrahams commented:
on Thu May 25 2017, Nate Cook <notifications-AT-github.com> wrote:

> The assumption is that a fixed width integer's `Magnitude` will always
> have the same `bitWidth`as the type itself. 
> The assumption is explicit with `DoubleWidth`, since it uses
> `Magnitude` as the lower half.

That relationship needn't be forced, though I suppose it would
complicate overflow detection if that were not the case.  In principle,
the compiler should be able to optimize that complication away when the
`bitWidth`'s do match, but that's a lot of code complexity for no gain in
practice.

I think it would be fine to make it a requirement on `DoubleWidth's`
parameter without pushing it into the protocol.

-- 
-Dave

dabrahams commented:
> the heterogeneous comparison with `0 as IntegerLiteralType` is less
> performant, 

Oh, that is sad.  Hopefully that will get fixed before we ship!

-- 
-Dave

xwu commented:
@dabrahams Well, it was a presumption based on reading the code. Maxim has pointed out that the difference is negligible when the type is `Int`, and it appears that's the case even with actual heterogeneous comparison, so I guess that's not a concern:

```swift
    @inline(never)
    func f<T : BinaryInteger>(_ x: T) -> Bool {
      return x != 0
    }
    var result = false
    measure {
      for i in 0..<1_000_000_000 {
        result = result || f(UInt16(i % Int(UInt16.max)))
      }
    } // 6.962 seconds
```

```swift
    @inline(never)
    func g<T : BinaryInteger>(_ x: T) -> Bool {
      return x != (0 as T)
    }
    var result = false
    measure {
      for i in 0..<1_000_000_000 {
        result = result || g(UInt16(i % Int(UInt16.max)))
      }
    } // 7.001 seconds
```

dabrahams commented:
Fortunately looking at the code isn't a good way to tell 😉 
If you inspect the SIL (or at least the generated asssmbly) they should be identical.

dabrahams commented:
On second thought, I'm wrong about the matching `bitWidth` requirement on `Magnitude`.  A big point of the `FixedWidthInteger` protocol is enabling the construction of multiple precision integers, and `DoubleWidth` is supposed to be a test case for that.  Obviously this means we need to add the requirement to  `FixedWidthInteger`!

natecook1000 committed:
Additional notes from @xwu

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
One more from @xwu + 80-column fixes

moiseev commented:
@natecook1000 are you happy with this one or would like to change some more?

natecook1000 commented:
@moiseev This is ready to go! 👍

moiseev commented:
@swift-ci Please Test Source Compatibility

moiseev commented:
@swift-ci Please test

xwu commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2361,8 +2395,7 @@ ${operatorComment('&' + x.operator, True)}
 /// An integer type that can represent only nonnegative values.
-public protocol UnsignedInteger : BinaryInteger
-  where Magnitude : BinaryInteger { }
+public protocol UnsignedInteger : BinaryInteger { }
```
@natecook1000 I didn't notice this change the first time round. Is there a reason that `UnsignedInteger` (and, below, `SignedInteger`) are losing the constraint `where Magnitude : BinaryInteger`?

moiseev replied:
I think it's now in `BinaryInteger` itself.

natecook1000 replied:
Yep! Constraints on associated types can be recursive as long as the associated type is declared in a higher protocol.

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -1339,8 +1339,9 @@ extension Numeric {
-  Hashable, Numeric, CustomStringConvertible, Strideable {
-
+  Hashable, Numeric, CustomStringConvertible, Strideable
+  where Magnitude : BinaryInteger, Magnitude.Magnitude == Magnitude
```
@xwu here

moiseev commented:
@swift-ci Please Test Source Compatibility
